Project: Language middleware

eatyourgreens opened the pull request:
`app-project` middleware which looks for a `language` query parameter and sets the locale.

This allows us to support links to FEM project pages from the PFE project navigation bar (see https://github.com/zooniverse/Panoptes-Front-End/pull/6146), from https://translations.zooniverse.org, and from the Translations page in the Project Builder.

## Package
app-project

## How to Review
Try it out with the test locale.
https://local.zooniverse.org:3000/projects/nora-dot-eisner/planet-hunters-tess/about/research?env=production&language=test

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [x] Tests are passing locally and on Github
- [x] Documentation is up to date and changelog has been updated if appropriate
- [x] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [x] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [x] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [x] Can submit a classification
- [x] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## New Feature
- [x] The PR creator has listed user actions to use when testing the new feature
- [ ] Unit tests are included for the new feature
- [ ] A storybook story has been created or updated
  - Example of SlideTutorial [component](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js) with docgen comments, and its [story](https://zooniverse.github.io/front-end-monorepo/@zooniverse/classifier/index.html?path=/docs/other-slidetutorial--default)

goplayoutside3 commented:
While this middleware does translate the page as expected, I think we should handle the ProjectSwitcher component before this strategy is merged.

I ran this PR locally with LocaleSwitcher rendered in ProjectHeader and visited the link in the PR description: https://local.zooniverse.org:3000/projects/nora-dot-eisner/planet-hunters-tess/about/research?env=production&language=test. Static strings are translated as expected. Then, I clicked "English" in the LocaleSwitcher and it takes me to: https://local.zooniverse.org:3000/projects/en/nora-dot-eisner/planet-hunters-tess/about/research?env=production&language=test. This is confusing because `/en/` isn't supposed to be used in FEM and the language query param is still `test`. Refreshing the page with the latter url causes the page to be translated with the test language.

1. Is there a way to fully redirect the browser from /projects/nora-dot-eisner/planet-hunters-tess?language=test to /projects/test/nora-dot-eisner/planet-hunters-tess?  or...
2. Should there be a handler when LocaleSwitcher changes to check if language query param exists and remove it from the url?

eatyourgreens commented:
There’s a missing line in this PR. It should delete the query parameter. 

`url.searchparams.delete(‘language’)`

eatyourgreens commented:
With the latest changes to https://github.com/zooniverse/Panoptes-Front-End/pull/6146, this PR is probably only needed for any legacy links that still use the `?language=es` query param to load a project in a specific language.

eatyourgreens commented on packages/app-project/pages/_middleware.js:
```diff
@@ -0,0 +1,16 @@
+    const locale = searchParams.get('language')
+    const url = req.nextUrl.clone()
+    try {
+      url.locale = locale
```
Somewhere around here, the old language query param should be deleted from the new URL. NextJS uses the standard [URLSearchParams API](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) so the syntax is something like:

```js
url.searchParams.delete('language')
```

goplayoutside3 commented:
> I ran this PR locally with LocaleSwitcher rendered in ProjectHeader and visited the link in the PR description: https://local.zooniverse.org:3000/projects/nora-dot-eisner/planet-hunters-tess/about/research?env=production&language=test. Static strings are translated as expected. Then, I clicked "English" in the LocaleSwitcher and it takes me to: https://local.zooniverse.org:3000/projects/en/nora-dot-eisner/planet-hunters-tess/about/research?env=production&language=test. This is confusing because `/en/` isn't supposed to be used in FEM and the language query param is still `test`. Refreshing the page with the latter url causes the page to be translated with the test language.

Wanted to check in and see if further work is planned on this PR. I re-tested the same behavior mentioned above, and I still see the unclear url where `/en/` appears and the `language` query param isn’t removed after using the LocaleSwitcher to select English.

eatyourgreens commented:
I'm catching up with this PR now. Trying to remember where it had got to.

eatyourgreens commented:
Hmmm. When I delete the `language` parameter (`url.searchParams.delete('language')`) I run into the following NextJS error: [Deleting query parameters in middlewares](https://nextjs.org/docs/messages/deleting-query-params-in-middlewares). I'm not sure how to fix that.

eatyourgreens commented:
Maybe this should redirect to a new URL, rather than rewriting the incoming request.

eatyourgreens commented:
@goplayoutside3 I've rewritten this to redirect, rather than rewrite, the URL. I've also added checks so that it doesn't redirect data requests, assets or the server health check page. It should all be working now: a URL like `/projects/owner/name?env=staging&language=fr` should redirect to `/fr/projects/owner/name?env=staging`

goplayoutside3 reviewed (approved):
Looks good!
